src/mqt/ddsim/primitives/estimator.py

```diff
@@ -10,230 +10,125 @@
-        """
-        state_circuits = []
-        observable_circuits = []
-        for group in self._grouping:
-            circuit = self._circuits[group[0]]
-            observable = self._observables[group[1]]
-            diff_circuits: list[QuantumCircuit] = []
-            if self._abelian_grouping:
-                for obs in observable.group_commuting(qubit_wise=True):
-                    basis = Pauli((
-                        np.logical_or.reduce(obs.paulis.z),
-                        np.logical_or.reduce(obs.paulis.x),
-                    ))
-                    obs_circuit, indices = self._observable_circuit(circuit.num_qubits, basis)
-                    paulis = PauliList.from_symplectic(
-                        obs.paulis.z[:, indices],
-                        obs.paulis.x[:, indices],
-                        obs.paulis.phase,
-                    )
-                    obs_circuit.metadata = {
-                        "paulis": paulis,
-                        "coeffs": np.real_if_close(obs.coeffs),
-                    }
-                    diff_circuits.append(obs_circuit)
-            else:
-                for basis, obs in zip(observable.paulis, observable):
-                    obs_circuit, indices = self._observable_circuit(circuit.num_qubits, basis)
-                    paulis = PauliList.from_symplectic(
-                        obs.paulis.z[:, indices],
-                        obs.paulis.x[:, indices],
-                        obs.paulis.phase,
-                    )
-                    obs_circuit.metadata = {
-                        "paulis": paulis,
-                        "coeffs": np.real_if_close(obs.coeffs),
-                    }
-                    diff_circuits.append(obs_circuit)
-
-            state_circuits.append(circuit.copy())
-            observable_circuits.append(diff_circuits)
-        return state_circuits, observable_circuits
+            observable_circuits[index] = (coeffs, observable_circuits_list)
 
-    @staticmethod
-    def _observable_circuit(num_qubits: int, pauli: Pauli) -> tuple[QuantumCircuit, list[int]]:
-        """Creates the quantum circuit representation of an observable given as a Pauli string.
-
-        Parameters:
-            - num_qubits: Number of qubits of the observable.
-            - pauli_string (str): The Pauli string representing the observable.
+        return observable_circuits
 
-        Returns:
-        Tuple: A tuple containing two entries:
-            - QuantumCircuit: The quantum circuit representation of the observable.
-            - List: A list of the qubits involved in the observables operators
-        """
+    @staticmethod
+    def _get_qubit_indices(pauli: Pauli) -> list[int]:
+        """Get the indices of the qubits that are part of the Pauli observable."""
         qubit_indices = np.arange(pauli.num_qubits)[pauli.z | pauli.x]
+        assert isinstance(qubit_indices, np.ndarray)
```

Comment:
Assertions like these should not actually be necessary as they are checked by mypy/static type checking.
I'd generally prefer to not have them in code, if they are not serving some real purpose.